Add new API so it's possible to not leak a key

woody-apple opened the pull request:
for CFTypes we need copy in front of them, otherwise it leaks

woody-apple committed:
Initial checkin

woody-apple committed:
Fixing review feedback

woody-apple committed:
Adding braces

restyled-commits committed:
Restyled by clang-format

kiel-apple commented on src/darwin/Framework/CHIP/MTRKeypair.h:
```diff
@@ -33,9 +33,15 @@ NS_ASSUME_NONNULL_BEGIN
+ * @brief Return public key for the keypair. DEPRECATED
+ */
+
+@property (nonatomic, readonly, copy) SecKeyRef publicKey MTR_DEPRECATED("Please implement copyPublicKey", ios(16.4, 17.2), macos(13.3, 14.2), watchos(9.4, 10.2), tvos(16.4, 17.2));
```
```suggestion
@property (nonatomic, readonly, copy) SecKeyRef publicKey MTR_DEPRECATED("Please use copyPublicKey instead.", ios(16.4, 17.2), macos(13.3, 14.2), watchos(9.4, 10.2), tvos(16.4, 17.2));
```

kiel-apple commented on examples/darwin-framework-tool/commands/common/CHIPToolKeypair.h:
```diff
@@ -22,7 +22,7 @@
 - (BOOL)initialize;
 - (NSData *)signMessageECDSA_RAW:(NSData *)message;
-- (SecKeyRef)publicKey;
+- (SecKeyRef)copyPublicKey;
```
```suggestion
- (SecKeyRef)publicKeyCopy;
```

kiel-apple commented on examples/darwin-framework-tool/commands/common/CHIPToolKeypair.mm:
```diff
@@ -65,7 +65,7 @@ - (NSData *)signMessageECDSA_RAW:(NSData *)message
 }
 
-- (SecKeyRef)publicKey
+- (SecKeyRef)copyPublicKey
```
```suggestion
- (SecKeyRef)publicKeyCopy
```

kiel-apple commented on src/darwin/Framework/CHIP/MTRCertificates.mm:
```diff
@@ -152,7 +152,24 @@ + (MTRCertificateDERBytes _Nullable)createOperationalCertificate:(id<MTRKeypair>
-    CHIP_ERROR err = MTRP256KeypairBridge::MatterPubKeyFromSecKeyRef(keypair.publicKey, &keypairPubKey);
+    SecKeyRef publicKey = NULL;
+
+    if ([keypair respondsToSelector:@selector(copyPublicKey)]) {
```
```suggestion
    if ([keypair respondsToSelector:@selector(publicKeyCopy)]) {
```

kiel-apple commented on src/darwin/Framework/CHIP/MTRCertificates.mm:
```diff
@@ -152,7 +152,24 @@ + (MTRCertificateDERBytes _Nullable)createOperationalCertificate:(id<MTRKeypair>
+    SecKeyRef publicKey = NULL;
+
+    if ([keypair respondsToSelector:@selector(copyPublicKey)]) {
+        publicKey = [keypair copyPublicKey];
```
```suggestion
        publicKey = [keypair publicKeyCopy];
```

kiel-apple commented on src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm:
```diff
@@ -819,7 +819,24 @@ - (BOOL)findMatchingFabric:(FabricTable &)fabricTable
-        CHIP_ERROR err = MTRP256KeypairBridge::MatterPubKeyFromSecKeyRef(params.nocSigner.publicKey, &pubKey);
+        SecKeyRef publicKey = NULL;
+
+        if ([params.nocSigner respondsToSelector:@selector(copyPublicKey)]) {
```
```suggestion
        if ([params.nocSigner respondsToSelector:@selector(publicKeyCopy)]) {
```

kiel-apple commented on src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm:
```diff
@@ -819,7 +819,24 @@ - (BOOL)findMatchingFabric:(FabricTable &)fabricTable
+        SecKeyRef publicKey = NULL;
+
+        if ([params.nocSigner respondsToSelector:@selector(copyPublicKey)]) {
+            publicKey = [params.nocSigner copyPublicKey];
```
```suggestion
            publicKey = [params.nocSigner publicKeyCopy];
```

kiel-apple commented on src/darwin/Framework/CHIP/MTRKeypair.h:
```diff
@@ -33,9 +33,15 @@ NS_ASSUME_NONNULL_BEGIN
+ * @brief Return a copy of the public key for the keypair.
  */
-- (SecKeyRef)publicKey;
+@property (nonatomic, readonly, copy) SecKeyRef copyPublicKey MTR_NEWLY_AVAILABLE;
```
```suggestion
@property (nonatomic, readonly, copy) SecKeyRef publicKeyCopy MTR_NEWLY_AVAILABLE;
```

kiel-apple commented on src/darwin/Framework/CHIPTests/MTRCertificateTests.m:
```diff
@@ -129,7 +129,7 @@ - (void)testGenerateIntermediateCert
     __auto_type * intermediateCert = [MTRCertificates createIntermediateCertificate:rootKeys
                                                                     rootCertificate:rootCert
-                                                              intermediatePublicKey:intermediateKeys.publicKey
+                                                              intermediatePublicKey:[intermediateKeys.copyPublicKey autorelease]
```
```suggestion
                                                              intermediatePublicKey:[intermediateKeys.publicKeyCopy autorelease]
```

kiel-apple commented:
ignore the comments - going to just do them all in a commit and see if that works better

woody-apple committed:
Fixing build

restyled-commits committed:
Restyled by clang-format

woody-apple committed:
Fixing annotations

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRKeypair.h:
```diff
@@ -31,13 +32,19 @@ NS_ASSUME_NONNULL_BEGIN
+- (SecKeyRef)copyPublicKey MTR_NEWLY_AVAILABLE;
+
+/**
+ * @brief Returns public key for the keypair. DEPRECATED - please use copyPublicKey, otherwise this will leak
```
```suggestion
 * @brief Returns public key for the keypair without adding a reference. DEPRECATED - please use copyPublicKey, otherwise this will leak.
```

bzbarsky-apple reviewed (commented):
Followup: `src/darwin/CHIPTool/CHIPTool/Framework Helpers/FabricKeys.m` needs fixing too.

woody-apple committed:
Update src/darwin/Framework/CHIP/MTRKeypair.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRKeypair.h:
```diff
@@ -31,13 +32,19 @@ NS_ASSUME_NONNULL_BEGIN
+ * @brief Returns public key for the keypair without adding a reference. DEPRECATED - please use copyPublicKey, otherwise this will leak.
+ */
+
+- (SecKeyRef)publicKey MTR_DEPRECATED("Please implement copyPublicKey, this will leak otherwise", ios(16.4, 18.3), macos(13.3, 15.3), watchos(9.4, 11.3), tvos(16.4, 18.3));
```
```suggestion
- (SecKeyRef)publicKey MTR_DEPRECATED("Please implement copyPublicKey, this will leak otherwise", ios(16.1, 18.3), macos(13.0, 15.3), watchos(9.1, 11.3), tvos(16.1, 18.3));
```
because I am 99% sure this dates back to initial release.  In particular, MTRDeviceControllerStartupParams is `MTR_AVAILABLE(ios(16.1), macos(13.0), watchos(9.1), tvos(16.1))` and has `id<MTRKeypair> nocSigner` with no other availability annotations....

kiel-apple commented on examples/darwin-framework-tool/commands/common/CHIPToolKeypair.mm:
```diff
@@ -79,7 +79,12 @@ - (SecKeyRef)publicKey
-    return _mPublicKey;
+
+    if (_mPublicKey) {
+        return CFRetain(_mPublicKey);
```
```suggestion
        CFRetain(_mPublicKey);
        return _mPublicKey;
```
avoids complaint from compiler when building `darwin-framework-tool`: `error: cannot initialize return object of type 'SecKeyRef' (aka '__SecKey *') with an rvalue of type 'CFTypeRef' (aka 'const void *')`

kiel-apple committed:
separate statements involving _mPublicKey to make compiler happy

kiel-apple committed:
use `CFAutorelease` on CoreFoundation typed public key copies

kiel-apple committed:
fix indent

kiel-apple committed:
implement `copyPublicKey` for `MTRTestKeys`; add TODO note about optional method calls

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRCertificateTests.m:
```diff
@@ -126,10 +126,16 @@ - (void)testGenerateIntermediateCert
 
     __auto_type * intermediateKeys = [[MTRTestKeys alloc] init];
     XCTAssertNotNil(intermediateKeys);
+    // TODO:  there are several places in our tests where we call `copyPublicKey` on objects without first
```
Yes, but we know we're using MTRTestKeys here, and that it implements copyPublicKey.  I don't think we need this comment.

kiel-apple replied:
my ambition was to make "there's no implementation of `copyPublicKey`" a compile-time error, but it looks like if it's optional in the protocol, that's not going to happen, so agreed and i'll take the comment out.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/TestHelpers/MTRTestKeys.m:
```diff
@@ -31,6 +31,13 @@ - (NSData *)publicKeyData
 }
 
+- (SecKeyRef)copyPublicKey {
+    // because this varies significantly from the actual implementations,
+    // it's probably not a great idea to rely on this for memory leak behavior information.
```
I don't get this comment.  The implementation here ("I have a key, and I will CFRetain it and return it to you") is _exactly_ what a "normal" MTRKeyPair implementation will do.

kiel-apple committed:
remove comment

it's a test; this is the best we can do with an optional protocol method

kiel-apple committed:
consistent formatting for `copyPublicKey` calls

kiel-apple committed:
reformat `copyPublicKey` in `MTRTestKeys`

remove comment - not this method's job to worry about other implementation's potential side-effects

woody-apple committed:
Update src/darwin/Framework/CHIP/MTRKeypair.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

kiel-apple commented:
re-rolling in #36430
